hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java

```diff
@@ -310,30 +339,67 @@ public static CryptoTransferWrapper decodeCryptoTransfer(
         final List<TokenTransferWrapper> tokenTransferWrappers = new ArrayList<>();
 
         for (final var tuple : decodedTuples) {
-            for (final var tupleNested : (Tuple[]) tuple) {
-                final var tokenType = convertAddressBytesToTokenID(tupleNested.get(0));
+            decodeTokenTransfer(aliasResolver, tokenTransferWrappers, (Tuple[]) tuple);
+        }
 
-                var nftExchanges = NO_NFT_EXCHANGES;
-                var fungibleTransfers = NO_FUNGIBLE_TRANSFERS;
+        return new CryptoTransferWrapper(new TransferWrapper(hbarTransfers), tokenTransferWrappers);
+    }
 
-                final var abiAdjustments = (Tuple[]) tupleNested.get(1);
-                if (abiAdjustments.length > 0) {
-                    fungibleTransfers =
-                            bindFungibleTransfersFrom(tokenType, abiAdjustments, aliasResolver);
-                }
-                final var abiNftExchanges = (Tuple[]) tupleNested.get(2);
-                if (abiNftExchanges.length > 0) {
-                    nftExchanges = bindNftExchangesFrom(tokenType, abiNftExchanges, aliasResolver);
-                }
+    /**
+     * Decodes the given bytes of the cryptoTransfer function parameters
+     *
+     * <p><b>Important: </b>This is the latest version and supersedes public static
+     * CryptoTransferWrapper decodeCryptoTransfer(). The selector for this function is derived from:
+     * cryptoTransfer(((address,int64,bool)[]),(address,(address,int64,bool)[],(address,address,int64,bool)[])[])
+     * The first parameter describes hbar transfers and the second describes token transfers
+     *
+     * @param input encoded bytes containing selector and input parameters
+     * @param aliasResolver function used to resolve aliases
+     * @return CryptoTransferWrapper codec
+     */
+    public static CryptoTransferWrapper decodeCryptoTransferV2(
+            final Bytes input, final UnaryOperator<byte[]> aliasResolver) {
+        final Tuple decodedTuples =
+                decodeFunctionCall(input, CRYPTO_TRANSFER_SELECTOR_V2, CRYPTO_TRANSFER_DECODER_V2);
+        List<SyntheticTxnFactory.HbarTransfer> hbarTransfers = new ArrayList<>();
+        final List<TokenTransferWrapper> tokenTransferWrappers = new ArrayList<>();
 
-                tokenTransferWrappers.add(
-                        new TokenTransferWrapper(nftExchanges, fungibleTransfers));
-            }
+        final Tuple[] hbarTransferTuples = ((Tuple) decodedTuples.get(0)).get(0);
+        final var tokenTransferTuples = decodedTuples.get(1);
+
+        if (hbarTransferTuples.length > 0) {
```

Comment:
nit: would be nice to have a `decodeHBarTransfer` to contain the appropriate logic

Reply:
done